libgadget/treewalk.c

```diff
@@ -281,47 +275,74 @@ cmpint(const void *a, const void *b)
+    {
+        const int p_i = active_set ? active_set[i] : i;
+        P[p_i].Evaluated = 0;
+    }
+}
+
+static void
+treewalk_build_queue(TreeWalk * tw, int * active_set, const int size, int may_have_garbage) {
+    int i;
+
+    if(!tw->haswork && !may_have_garbage)
+    {
+        tw->WorkSetSize = size;
+        tw->WorkSet = active_set;
+        tw->work_set_allocated = 0;
+        return;
+    }
+
+    tw->WorkSet = mymalloc("ActiveQueue", size * sizeof(int));
+    tw->work_set_allocated = 1;
+
     int * queue = tw->WorkSet;
-    int k = 0;
-    if(active_set == NULL) {
-        int i;
-        #pragma omp parallel for
-        for(i = 0; i < PartManager->NumPart; i++) {
-            /* Skip the garbage particles */
-            if(P[i].IsGarbage) continue;
-            if(!tw->haswork(i, tw))
-                continue;
-            const int lock = atomic_fetch_and_add(&k, 1);
-            queue[lock] = i;
-        }
-    } else {
-        int i;
-        #pragma omp parallel for
-        for(i=0; i < NumActiveParticle; i++)
-        {
-            const int p_i = ActiveParticle[i];
+    int nqueue = 0;
 
-            /* Skip the garbage particles */
-            if(P[p_i].IsGarbage) continue;
+    /*We want a lockless algorithm which preserves the ordering of the particle list.*/
+    int nqthr[All.NumThreads];
+    memset(nqthr, 0, All.NumThreads*sizeof(int));
+    int * thrqueue = mymalloc("threadedqueue", size * sizeof(int) * All.NumThreads);
+
+    /* We enforce schedule static to ensure that each thread executes on contiguous particles.*/
+    #pragma omp parallel for schedule(static)
+    for(i=0; i < size; i++)
+    {
+        const int tid = omp_get_thread_num();
+        /*Use raw particle number if active_set is null, otherwise use active_set*/
+        const int p_i = active_set ? active_set[i] : i;
+
+        /* Skip the garbage particles */
+        if(P[p_i].IsGarbage) continue;
+
+        if(tw->haswork && !tw->haswork(p_i, tw))
+            continue;
+        thrqueue[size * tid + nqthr[tid]] = p_i;
+        nqthr[tid]++;
+    }
+    /*Merge step for the queue.*/
+    nqueue = 0;
+    for(i = 0; i < All.NumThreads; i++)
+    {
+        memmove(queue + nqueue, thrqueue + i*size, sizeof(int) * nqthr[i]);
+        nqueue += nqthr[i];
+    }
+    myfree(thrqueue);
 
-            if(!tw->haswork(p_i, tw))
-               continue;
-            const int lock = atomic_fetch_and_add(&k, 1);
-            queue[lock] = p_i;
-        }
 #ifdef DEBUG
-        /* check the uniqueness of ActiveParticle list. */
-        /* FIXME: the sort may affect performance of treewalk */
-        qsort_openmp(queue, k, sizeof(int), cmpint);
-        for(i = 0; i < k - 1; i ++) {
-            if(queue[i] == queue[i+1]) {
-                endrun(8829, "A few particles are twicely active.");
-            }
+    /* check the uniqueness of the active_set list. This is very slow. */
+    qsort_openmp(queue, nqueue, sizeof(int), cmpint);
```

Comment:
We can replace this with a hash table of length nqueue and type char. Then it is blazingly fast.

Reply:
Sure - Do we have a hash table structure around? This is within a DEBUG ifdef, so the speed is not vital.